app/app.go

```diff
@@ -495,3 +495,57 @@ func (a *App) Handle404(w http.ResponseWriter, r *http.Request) {
 
 	utils.RenderWebAppError(w, r, err, a.AsymmetricSigningKey())
 }
+
+// This function migrates the default built in roles from code/config to the database.
+func (a *App) DoAdvancedPermissionsMigration() {
+	// If the migration is already marked as completed, don't do it again.
+	if result := <-a.Srv.Store.System().GetByName(ADVANCED_PERMISSIONS_MIGRATION_KEY); result.Err == nil {
```

Comment:
Any reason not to just leverage a transaction? Or is this path expose to explicitly allow someone to clear the bit and force the migration a second time?

Reply:
(This question implicitly assumes it makes more sense to do in `upgrade.go`, but you've answered that question. I guess my only question is whether or not it still makes sense to do it as an all-or-nothing transaction, but I do note that it seems to self-recover if a role already exists.)

Reply:
Context for this one is here: https://github.com/mattermost/mattermost-server/pull/8159/files/a14b410200182a567c053278048b1745462bcebc..ed80e66b72b7fcff4cc3ae64f15d4da9ddedb034#diff-886cebc00c073731af1b3da26f11e536R468